File: src/main/java/programmers/Lim84512.java
package programmers;

public class Lim84512 {
    int answer = 0;
    boolean solved = false;

    public int solution(String word) {
        final String[] vowels = {"A", "E", "I", "O", "U"};
        for (int i = 0; i < vowels.length; i++) {
            solve(i, "", vowels, word, 1);
        }

        return answer;
    }

    private void solve(int idx, String value, String[] vowels, String word, int depth) {
        if (solved) {
            return;
        }
        answer++;
        final String result = value + vowels[idx];
        if (result.equals(word)) {
            solved = true;
            return;
        }
        if (depth == 5) {
            return;
        }

        for (int i = 0; i < vowels.length; i++) {
            solve(i, result, vowels, word, depth + 1);
        }
    }
}

File: src/main/java/programmers/part2/Lim84512.kt
package programmers.part2

class Lim84512 {
    private val vowels = listOf('A', 'E', 'I', 'O', 'U')
    private var answer = 0
    private var solved = false

    fun solution(word: String): Int {
        for (i in vowels.indices) {
            dfs(i, emptyList(), word, 1)
        }

        return answer
    }

    private fun dfs(idx: Int, x: List<Char>, word: String, depth: Int) {
        if (solved.not()) {
            answer++
            val next = x + vowels[idx]

            if (next.joinToString("") == word) {
                solved = true
                return
            }
            if (depth == 5) {
                return
            }

            for (i in vowels.indices) {
                dfs(i, next, word, depth + 1)
            }
        }
    }
}
